lib/middleware/early_hints.rb

```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+require_dependency "mobile_detection"
+
+module Middleware
+  class EarlyHints
+    def initialize(app)
+      @app = app
+    end
+
+    def call(env)
+      @env = env
+      @request = ActionDispatch::Request.new(@env)
+      if hints_useful?
+        @request.send_early_hints("Link" => links)
+      end
+      @app.call(@env)
+    end
+
+    private
+
+    # Not compreensive but still useful
+    def hints_useful?
+      @request.get? &&
+      !@request.xhr? &&
+      !@request.path.ends_with?('robots.txt') &&
+      !@request.path.ends_with?('srv/status') &&
+      @request[Auth::DefaultCurrentUserProvider::API_KEY].nil?
+    end
+
+    def links
+      unless hints = Discourse.redis.get(hints_cache_key)
+        hints = hints_list
+        Discourse.redis.setex(hints_cache_key, hints_cache_duration, hints)
+      end
+
+      hints
+    end
+
+    def hints_cache_key
+      "discourse-early-hints-cache-#{is_mobile? ? 'mobile' : 'desktop'}"
+    end
+
+    def hints_cache_duration
+      5 * 60
+    end
+
+    def hints_list
+      js_hints = %w(
+        application.js
+        vendor.js
+      )
+
+      js_hints.map do |asset|
+        hint_builder(
+          UrlHelper.absolute("/assets/#{manifest_asset(asset)}"),
+          'script'
+        )
+      end.append(
+        hint_builder(
+          main_css_file, "style"
+        )
+      ).join(',')
+    end
+
+    def hint_builder(url, type)
+      "<#{url}>; rel=\"preload\"; as=\"#{type}\""
+    end
+
+    def manifest_asset(asset)
+      Rails.application.assets_manifest.files.detect { |k, v| v['logical_path'] == asset }.first
```

Comment:
On my tests the browser was downloading the brotli version by itself, as it does automagically when we ask for ` <script src="/main.js">`  in a HTML doc.

Reply:
Ohhh I see the hack we added to bypass CDN lack of brotli support 5 years ago is still biting us. It may be wise to drop this hack eventually.

Anyway, since nginx doesn't support early hints this feature flag is not something most sites (our hosting included) will be able to enable any time soon. Sites would need to use either HAProxy or H20 *inside* the container in order to enable this.

So maybe I just close this PR?

Reply:
> nginx doesn't support early hints

oooh that is annoying! Then yeah I guess we either close it, or add some big 'experimental, does not work with NGINX, or CDN/S3 assets' comment to the top of the file.